File: algorithm_introduction/ss6_heapsort/python/max_priority_queue.py
from collections import UserList
from typing import List


class MaxPriorityQueue(UserList):
    def __init__(self, initializer: List[int]):
        self.data = initializer
        self.last_heap_idx = len(initializer) - 1
        self._build_max_heap()

    def __str__(self) -> str:
        return str(self.data[:self.last_heap_idx + 1])

    def _parent(self, i: int):
        return (i - 1) >> 1

    def _left(self, i: int):
        return ((i + 1) << 1) - 1

    def _right(self, i: int):
        return (i + 1) << 1

    def _max_heapify(self, i: int):
        largest = i
        l = self._left(i)
        r = self._right(i)
        if l <= self.last_heap_idx and self[l] > self[largest]:
            largest = l
        if r <= self.last_heap_idx and self[r] > self[largest]:
            largest = r
        if largest != i:
            self[i], self[largest] = self[largest], self[i]
            self._max_heapify(largest)

    def _build_max_heap(self):
        for i in range((len(self) >> 1) - 1, -1, -1):
            self._max_heapify(i)

    def heap_maximum(self):
        return self[0]

    def heap_extract_max(self):
        if self.last_heap_idx < 0:
            raise ValueError('Heap underflow')
        max_val = self[0]
        self[0] = self[self.last_heap_idx]
        self.last_heap_idx -= 1
        self._max_heapify(0)
        return max_val

    def _heap_increase_key(self, i: int, key: int):
        if key < self[i]:
            raise ValueError(f'New key (={key}) must be greater than current key (={self[i]}).')
        self[i] = key
        while i > 0 and self[self._parent(i)] < self[i]:
            self[self._parent(i)], self[i] = self[i], self[self._parent(i)]
            i = self._parent(i)

    def max_heap_insert(self, key: int):
        self.last_heap_idx += 1
        self.append(-float('inf'))
        self._heap_increase_key(self.last_heap_idx, key)


def main():
    cases = {
        'A': MaxPriorityQueue([4, 1, 3, 2, 16, 9, 10, 14, 8, 7]),
        'B': MaxPriorityQueue([4, 1, 3, 2, 16, 9, 10, 14, 8]),
    }

    for k, v in cases.items():
        print(f'# {k}')
        print(v)
        print(v.heap_extract_max())
        print(v)
        print(v.heap_extract_max())
        print(v)
        print(v.heap_extract_max())
        print(v)
        print(v.heap_extract_max())
        print(v)
        v.max_heap_insert(100)
        print(v)
        v.max_heap_insert(6)
        print(v)
        print(v.heap_extract_max())
        print(v)
        print(v.heap_extract_max())
        print(v)
        print()


if __name__ == '__main__':
    main()

File: algorithm_introduction/ss06_heapsort/cpp/max_priority_queue.cpp
#include <iostream>
#include <limits>
#include <string>
#include <unordered_map>

using namespace std;

struct MaxPriorityQueue {
  int *arr;
  int capacity;
  int last_heap_idx;

  string display() {
    string s = "[";
    for (int i = 0; i <= last_heap_idx; ++i) {
      s += to_string(arr[i]);
      s += ", ";
    }
    s += "]";
    return s;
  }

  MaxPriorityQueue(int *ini_arr, const int &ini_capacity, int ini_last_heap_idx)
      : arr(ini_arr), capacity(ini_capacity), last_heap_idx(ini_last_heap_idx) {
    _build_max_heap();
  }

  int _parent(int i) { return (i - 1) >> 1; }

  int _left(int i) { return ((i + 1) << 1) - 1; }

  int _right(int i) { return (i + 1) << 1; }

  void _max_heapify(int i) {
    int largest = i;
    int l = _left(i);
    int r = _right(i);
    if (l <= last_heap_idx && arr[l] > arr[largest]) {
      largest = l;
    }
    if (r <= last_heap_idx && arr[r] > arr[largest]) {
      largest = r;
    }
    if (largest != i) {
      swap(arr[i], arr[largest]);
      _max_heapify(largest);
    }
  }

  void _build_max_heap() {
    for (int i = _parent(last_heap_idx); i >= 0; --i) {
      _max_heapify(i);
    }
  }

  int heap_maximum() { return arr[0]; }

  int heap_extract_max() {
    if (last_heap_idx < 0) {
      throw "Heap underflow";
    }
    int max_val = arr[0];
    arr[0] = arr[last_heap_idx];
    last_heap_idx -= 1;
    _max_heapify(0);
    return max_val;
  }

  void _heap_increase_key(int i, int key) {
    if (key < arr[i]) {
      throw "New key (=" + to_string(key) +
          ") must be greater than current key (=" + to_string(arr[i]) + ".)";
    }
    arr[i] = key;
    while (i > 0 && arr[_parent(i)] < arr[i]) {
      swap(arr[_parent(i)], arr[i]);
      i = _parent(i);
    }
  }

  void max_heap_insert(int key) {
    last_heap_idx += 1;
    arr[last_heap_idx] = numeric_limits<int>::lowest();
    _heap_increase_key(last_heap_idx, key);
  }
};

int main() {
  int A[] = {4, 1, 3, 2, 16, 9, 10, 14, 8, 7};
  int B[] = {4, 1, 3, 2, 16, 9, 10, 14, 8};
  MaxPriorityQueue que_A(A, 100, (int)(sizeof(A) / sizeof(A[0]) - 1));
  MaxPriorityQueue que_B(B, 100, (int)(sizeof(B) / sizeof(B[0]) - 1));
  unordered_map<string, MaxPriorityQueue *> cases{{"A", &que_A}, {"B", &que_B}};

  for (auto &p : cases) {
    cout << "# " << p.first << "\n";
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    p.second->max_heap_insert(100);
    cout << p.second->display() << "\n";
    p.second->max_heap_insert(6);
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    cout << to_string(p.second->heap_extract_max()) << "\n";
    cout << p.second->display() << "\n";
    cout << endl;
  }
}